### PR TITLE
Automatically run ansible-lint on each role

### PR DESCRIPTION
**Pull request by ajdecon**
## Summary

This PR makes the following changes:

- Adds `scripts/deepops/ansible-lint-roles.sh` which runs `ansible-lint` on each role in the `roles/` directory
- Adds a Github action in `.github/workflows/ansible-lint-roles.yml` to run this automatically on pushes and PRs
- Adds documentation on these scripts to `docs/deepops/testing.md`
- Cleans up linting issues in most existing roles so that the script can pass

(This turned out to be less difficult than I thought! Most roles had a small number of issues, or one minor issue repeated many times. A lot of fixes were solvable with a clever regex or three.)

Note: `ansible-lint-roles.sh` includes the option of specifying an environment variable `ANSIBLE_LINT_EXCLUDE`, that will be used as a regex to skip linting some roles. This is currently used in the Github action to skip linting `nvidia-dgx`, `nvidia-dgx-firmware`, and `nvidia-gpu-tests`. The first role is pending major changes in #1117. The other two are complex enough, and difficult enough to test, that they should get extra testing when making the changes needed for linting to pass.

## Test plan

- [x] All roles should pass `ansible-lint`, unless specifically excluded
- [x] All existing Ansible Molecule tests should pass
- [x] All existing end-to-end Jenkins tests should pass

### CODE REVIEW

**Comment by supertetelman** [comment]
Wow, 76 files edited?

Is there any specific area you think this could be potentially risky that I should focus on with a review?

File: docs/deepops/testing.md [hunk 1]
```diff
@@ -1,145 +1,158 @@
...
-Additional services such as Kubeflow, Open OnDemand, or Monitoring may have additional validation steps that are documented in the corresponding DeepOps READMEs and the official documentation.
-
-For workloads that can be used as post-deployment validation, see the example workloads for [k8s]( ../../workloads/examples/k8s/) and [slurm]( ../../workloads/examples/slurm).
+# DeepOps Testing, CI/CD, and Validation
+
+## Linting
+
+`ansible-lint` is automatically run for each role in the `roles/` directory using a [Github action](../../.github/workflows/ansible-lint-roles.yml).
+This action runs `ansible-lint` for each role, and provides both the full output and a list of roles that failed linting.
+If the Github action reports success, all roles should have passed linting.
```
Comment (supertetelman) [inline docs/deepops/testing.md:7]: Github -> GitHub

File: docs/deepops/testing.md [hunk 2]
```diff
@@ -1,145 +1,158 @@
...
-The Slurm and Kubernetes deployment guides both document cluster verification steps. These should be run during the installation process to validate a GPU workload can be executed on the cluster.
-
-Additional services such as Kubeflow, Open OnDemand, or Monitoring may have additional validation steps that are documented in the corresponding DeepOps READMEs and the official documentation.
-
-For workloads that can be used as post-deployment validation, see the example workloads for [k8s]( ../../workloads/examples/k8s/) and [slurm]( ../../workloads/examples/slurm).
+# DeepOps Testing, CI/CD, and Validation
+
+## Linting
+
+`ansible-lint` is automatically run for each role in the `roles/` directory using a [Github action](../../.github/workflows/ansible-lint-roles.yml).
```
Comment (supertetelman) [inline docs/deepops/testing.md:5]: Github -> GitHub

File: roles/autofs/tasks/main.yml [hunk 3]
```diff
@@ -22,7 +22,12 @@
     - configuration
 
 - name: ensure mountpoint exists
-  file: path={{ autofs_mount }} state=directory
+  file: 
+    path: "{{ autofs_mount }}"
+    state: directory
+    owner: "root"
+    group: "root"
+    mode: "0755"
```
Comment (supertetelman) [inline roles/autofs/tasks/main.yml:30]: Is this mode value just the default? It does not appear in the previous main.yml

File: roles/cachefilesd/tasks/main.yml [hunk 4]
```diff
@@ -8,11 +8,17 @@
   template:
     src: cachefilesd.j2
     dest: /etc/default/cachefilesd
+    owner: "root"
+    group: "root"
+    mode: "0644"
```
Comment (supertetelman) [inline roles/cachefilesd/tasks/main.yml:13]: Where did these values come from?

File: roles/docker-rootless/tasks/main.yml [hunk 5]
```diff
@@ -61,8 +61,7 @@
 
 - name: Execute setup rootless docker script
   become: true
-  shell:
-    /tmp/rootless_docker_setup.sh
+  command: /tmp/rootless_docker_setup.sh
```
Comment (supertetelman) [inline roles/docker-rootless/tasks/main.yml:64]: Has this been tested? Sometimes shell is needed for parsing output, although that does not appear to be the case here.

File: roles/easy-build-packages/defaults/main.yml [hunk 6]
```diff
@@ -1 +1,2 @@
-sm_files_path: "{{ sm_prefix }}/easybuild_files"
\ No newline at end of file
+sm_files_path: "{{ sm_prefix }}/easybuild_files"
+sm_files_repo_version: "4ef7ae6cc2284f69412a8db5e10dddd92024eeab"
```
Comment (supertetelman) [inline roles/easy-build-packages/defaults/main.yml:2]: What's up with this repo_version?

File: roles/easy-build-packages/.ansible-lint [hunk 7]
```diff
@@ -0,0 +1,4 @@
+skip_list:
```
Comment (supertetelman) [inline roles/easy-build-packages/.ansible-lint:1]: Was this automatically added?

File: roles/easy-build/tasks/main.yml [hunk 8]
```diff
@@ -56,7 +56,7 @@
 
 - name: "get EasyBuild version"
   become: yes
-  shell: ls {{ eb_tmp_dir }}/software/EasyBuild/
+  command: ls {{ eb_tmp_dir }}/software/EasyBuild/
```
Comment (supertetelman) [inline roles/easy-build/tasks/main.yml:59]: Has this been tested?

File: roles/k8s-internal-container-registry/tasks/main.yml [hunk 9]
```diff
@@ -6,9 +6,11 @@
 
 - name: install stable helm repository
   command: "{{ helm_cmd }} repo add {{ container_registry_repo_name }} {{ container_registry_chart_location }}"
+  changed_when: false
```
Comment (supertetelman) [inline roles/k8s-internal-container-registry/tasks/main.yml:9]: Why did you add this? This should be consistent with all the helm install roles.

File: roles/lmod/tasks/main.yml [hunk 10]
```diff
@@ -73,46 +79,52 @@
...
+    src: "/usr/lib/x86_64-linux-gnu/lua/{{ item }}/posix_c.so"
+    dest: "/usr/lib/x86_64-linux-gnu/lua/{{ item }}/posix.so"
+    owner: "root"
+    group: "root"
+    state: "link"
+    force: true
   when: 
     - ansible_distribution == 'Ubuntu'
     - ansible_distribution_version == '18.04'
-  changed_when: lmod_bugfix_posix.rc != 0
```
Comment (supertetelman) [inline roles/lmod/tasks/main.yml:91]: Is it okay that we dropped this?

File: roles/nvidia-dgx-firmware/tasks/get-data.yml [hunk 11]
```diff
@@ -3,6 +3,6 @@
 - name: Run NVSM human-readable health show
   shell: "osversion=`cat /etc/dgx-release | grep 'DGX_SWBUILD_VERSION'`; host=`hostname`; copper_mac=`ip addr | grep -A 1 enp6s0 | grep link | awk '{print $2}'`; bmc_ip=`sudo ipmitool lan print | grep -i 'IP Address              :' | awk '{print $4}'`; bmc_mac=`sudo ipmitool lan print | grep -i 'MAC Address' | awk '{print $4}'`;host_mac=`ifconfig {{ nv_mgmt_interface }} | grep 'ether' | awk '{print $2}'`;host_ip=`ifconfig {{ nv_mgmt_interface }} | grep 'inet ' | awk '{print $2}'`;data=\"${host_ip},${host_mac},${bmc_ip},${bmc_mac},${copper_mac},${host},${osversion}\";echo ${data}"
   register: command_output
-  - debug:
```
Comment (supertetelman) [inline roles/nvidia-dgx-firmware/tasks/get-data.yml:6]: How was this not failing previously from indentation errors?

File: roles/nvidia-gpu-operator-node-prep/tasks/main.yml [hunk 12]
```diff
@@ -7,7 +7,7 @@
   modprobe:
     name: nouveau
     state: absent
-  ignore_errors: true
+  failed_when: false
```
Comment (supertetelman) [inline roles/nvidia-gpu-operator-node-prep/tasks/main.yml:10]: Any idea what the functional difference here is?

File: roles/nvidia-gpu-operator/tasks/k8s.yml [hunk 13]
```diff
@@ -5,12 +5,16 @@
...
 - name: install gpu-operator helm repo
   command: /usr/local/bin/helm repo add nvidia "{{ gpu_operator_helm_repo }}"
+  changed_when: false
 
 - name: update helm repos
   command: /usr/local/bin/helm repo update
+  changed_when: false
 
 - name: Remove existing GPU Operator labels to accommodate potenial configuration changes
-  shell: kubectl label node --all nvidia.com/gpu.deploy.container-toolkit-  nvidia.com/gpu.deploy.dcgm- nvidia.com/gpu.deploy.dcgm-exporter- nvidia.com/gpu.deploy.device-plugin- nvidia.com/gpu.deploy.driver- nvidia.com/gpu.deploy.gpu-feature-discovery- nvidia.com/gpu.deploy.node-status-exporter- nvidia.com/gpu.deploy.operator-validator-
```
Comment (supertetelman) [inline roles/nvidia-gpu-operator/tasks/k8s.yml:13]: I just added this line and if I remember there was an edge-case error that required me to use shell here instead of command. Let me double check that this is passing the tests.

File: roles/nvidia-gpu-operator/tasks/nvaie.yml [hunk 14]
```diff
@@ -5,7 +5,8 @@
     gpu_operator_enable_migmanager: false
 
 - name: Create namespace for GPU Operator resources
-  shell: kubectl create namespace {{ gpu_operator_namespace }} -o yaml --dry-run=client | kubectl apply -f -
+  shell: set -o pipefail && kubectl create namespace {{ gpu_operator_namespace }} -o yaml --dry-run=client | kubectl apply -f -
+  changed_when: false
```
Comment (supertetelman) [inline roles/nvidia-gpu-operator/tasks/nvaie.yml:9]: This should be be changed_when: true, shouldn't it?

File: scripts/deepops/ansible-lint-roles.sh [hunk 15]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
```
Comment (supertetelman) [inline scripts/deepops/ansible-lint-roles.sh:1]: Needs a line or 2 of comments documenting the usage of the script.

**Review by supertetelman — DISMISSED** [review]
Learned a few neat things from going through all this linting. I had 1 or 2 concerns for the changes on some of the less used/testable playbooks. I also question adding additional lines for "changed_when: False" for every single helm command, but I am assuming the official policy is to set that for any command or shell?

Otherwise LGTM and merge it in.

File: roles/nvidia-gpu-operator/tasks/k8s.yml (hunk 13 above)
Comment (supertetelman) [reply roles/nvidia-gpu-operator/tasks/k8s.yml:13]: Looks like no.

File: roles/lmod/tasks/main.yml (hunk 10 above)
Comment (ajdecon) [reply roles/lmod/tasks/main.yml:91]: Yes - this needed to be dropped because we aren't executing commands anymore, we are using the built-in file module. So there isn't even a `lmod_bugfix_posix.rc` variable to check anymore!

File: roles/nvidia-gpu-operator-node-prep/tasks/main.yml (hunk 12 above)
Comment (ajdecon) [reply roles/nvidia-gpu-operator-node-prep/tasks/main.yml:10]: AFAICT, the major difference is that the task isn't marked as `failed` in the final Ansible results summary.

**Comment by ajdecon** [comment]
@supertetelman : Thanks for walking through this! I definitely want good feedback on this PR, because it touches so many files. As you can see most of the changes are pretty mechanical, but that doesn't mean they're all perfect.

My major goal with this PR is really to get automated linting in place so that we can catch possible problems in the future, especially in code paths that aren't normally executed. The cleanups I did on the individual roles were intended to get us to an initial "passing" state to make the automated tests useful for future PRs.

Re: high-risk areas, I think anything not exercised by the end-to-end tests may be risky. I *think* most of the changes I made are pretty obviously benign. But I'm also happy to break risky roles out into their own PR for more thorough testing, and just add them to `ANSIBLE_EXCLUDE_LINT` for this PR.

A few high-level comments based on the things your highlighted:

- `ansible-lint` flags tasks that can create new files without specifying a `mode` parameter, because Ansible doesn't actually have a default value for this. 😬 Instead, it runs in the environment of whatever the `ansible_user` is, and uses the environment's `umask` to determine permissions.
    The mode I selected in most cases (`0644` for files, `0755` for directories) is a good general default based on the default umask for the root user on most distros (`022`).

- The `changed_when: false` additions to commands are being added for task idempotence purposes. This is a little bit of a hack -- ideally, commands would only be executed when needed. But where a command is *always* execution, the best practice I can find is to suppress the `changed` message, because that execution is the expected behavior on a "clean" fun.

- The `failed_when: false` replacements for `ignore_errors: true` is the most problematic change here. Ideally we'd be examining the output of each task and reporting failure based on parsing that. This PR was intended to be a "quick" addition of linting, so I just suppressed the expected failures completely... However, if you think it's a better idea, we can undo those changes. The presence of `ignore_errors: true` appears to be a linting warning rather than an error, so the overall lint process should still pass.

File: roles/nvidia-gpu-operator/tasks/nvaie.yml (hunk 14 above)
Comment (ajdecon) [reply roles/nvidia-gpu-operator/tasks/nvaie.yml:9]: See my high-level comment.

File: roles/easy-build-packages/defaults/main.yml (hunk 6 above)
Comment (ajdecon) [reply roles/easy-build-packages/defaults/main.yml:2]: Most recent commit in https://github.com/DeepOps/easyconfigs . `ansible-lint` doesn't like `HEAD` as a git checkout target.

File: roles/nvidia-gpu-operator-node-prep/tasks/main.yml (hunk 12 above)
Comment (supertetelman) [reply roles/nvidia-gpu-operator-node-prep/tasks/main.yml:10]: From what I can tell the difference is that ignore_errors makes the output red and failed_when doesn't.

File: roles/lmod/tasks/main.yml (hunk 10 above)
Comment (supertetelman) [reply roles/lmod/tasks/main.yml:91]: Makes sense!

File: roles/easy-build-packages/defaults/main.yml (hunk 6 above)
Comment (supertetelman) [reply roles/easy-build-packages/defaults/main.yml:2]: Okay, sounds good. I also prefer pointing to a specific hash.

File: roles/nvidia-mig-manager/tasks/main.yml [hunk 16]
```diff
@@ -3,13 +3,16 @@
...
 - name: check for MIG capable devices
-  shell: nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader | grep -v 'N/A'
+  shell: set -o pipefail && nvidia-smi --query-gpu=mig.mode.current --format=csv,noheader | grep -v 'N/A'
   register: has_mig
   failed_when: false
+  changed_when: false
+
 - name: check for NVIDIA MIG parted
-  shell: which nvidia-mig-parted
+  command: which nvidia-mig-parted
```
Comment (supertetelman) [inline roles/nvidia-mig-manager/tasks/main.yml:12]: I think this will break.

File: roles/roce_backend/tasks/mofed-install.yaml [hunk 17]
```diff
@@ -10,7 +11,7 @@
         path: /usr/bin/ofed_info
       register: ofed_installed
     - name: Check Mellanox OFED version
-      shell: /usr/bin/ofed_info -n
+      command: /usr/bin/ofed_info -n
```
Comment (supertetelman) [inline roles/roce_backend/tasks/mofed-install.yaml:14]: I'm not sure about this one.

**Comment by ajdecon** [comment]
@supertetelman : I think those are probably ok, but also those roles are tricky to test. So I've swapped those two cases back to `shell` and added `# noqa` comments.

File: scripts/deepops/ansible-lint-roles.sh (hunk 15 above)
Comment (ajdecon) [reply scripts/deepops/ansible-lint-roles.sh:1]: Done, see below.

**Review by supertetelman — APPROVED** [review]
Great work!
